### PR TITLE
Fix CSM Artefacts

### PR DESCRIPTION
**Pull request by johannesmayer**
For cartesian acquisition data that are undersampled, the reconstructions underlying the coil sensitivities computation can contain high-frequency artefacts that will propagate to the coil maps and hence to the resulting reconstructions.

The `Acquisition` objects' header, however, contain flags that label whether the acquisitions are suitable to be used for coil calibration. These flags are given by: 
```
ISMRMRD_ACQ_IS_PARALLEL_CALIBRATION             = 20
ISMRMRD_ACQ_IS_PARALLEL_CALIBRATION_AND_IMAGING = 21,
```

This PR uses only the subset of `Acquistions` that have these flags set for cartesian trajectories.
If no `Acquisition` contains these two flags, the entire set of raw data is employed in the coil sensitivity estimation.

### CODE REVIEW

File: examples/Python/MR/grappa_and_steepest_descent.py [hunk 1]
```diff
@@ -107,6 +107,9 @@ def main():
     acq_model = AcquisitionModel(preprocessed_data, image_data)
     acq_model.set_coil_sensitivity_maps(csms)
 
+    start_val = 0.001
+    image_data = image_data.fill(start_val + start_val * 1j)
```
Comment (KrisThielemans) [inline examples/Python/MR/grappa_and_steepest_descent.py:111]:
```suggestion
    image_data.fill(start_val + start_val * 1j)
```

File: src/xGadgetron/pGadgetron/Gadgetron.py [hunk 2]
```diff
@@ -1590,3 +1601,45 @@ def calc_rpe_dcw(ad):
...
     return dcw
+
+def extract_calibration_data(ad):
+    '''
+    Function to extract all Acquisitions from an AcquisitionData object that
+    carry the coil calibration flag.
+    '''
+
+    if not isinstance(ad, AcquisitionData):
+        raise AssertionError("The trajectory you asked for is not among the available trajectoryies")
```
Comment (KrisThielemans) [inline src/xGadgetron/pGadgetron/Gadgetron.py:1612]:
```suggestion
        raise AssertionError("The trajectory you asked for is not among the available trajectories")
```

File: src/xGadgetron/pGadgetron/Gadgetron.py [hunk 3]
```diff
@@ -1590,3 +1601,45 @@ def calc_rpe_dcw(ad):
...
+        acq = ad.acquisition(i)
+        flags = decode_ismrmrd_flag(acq.flags())
+        if ref_flag in flags or ref_img_flag in flags:
+            subset_idx = numpy.append(subset_idx, i)
+
+    if subset_idx.size > 0:
+        calib_data = ad.get_subset(subset_idx)
+        return calib_data
+    else:
+        print("Warning, there are no calibration data in this dataset. We use them all for CSM computation.")
```
Comment (KrisThielemans) [inline src/xGadgetron/pGadgetron/Gadgetron.py:1628]:
```suggestion
        print("Warning, there are no calibration data in this dataset. We use them for CSM computation.")
```
but ideally you add something about what the consequences are (if easy?)

File: src/xGadgetron/pGadgetron/Gadgetron.py [hunk 4]
```diff
@@ -1590,3 +1601,45 @@ def calc_rpe_dcw(ad):
...
+    '''
+    Function to extract all Acquisitions from an AcquisitionData object that
+    carry the coil calibration flag.
+    '''
+
+    if not isinstance(ad, AcquisitionData):
+        raise AssertionError("The trajectory you asked for is not among the available trajectoryies")
+
+    ref_flag = 20
+    ref_img_flag = 21
```
Comment (KrisThielemans) [inline src/xGadgetron/pGadgetron/Gadgetron.py:1615]: can you add some of the comments if what you wrote in the PR description here?

**Review by KrisThielemans — CHANGES_REQUESTED** [review]

**Review by evgueni-ovtchinnikov — CHANGES_REQUESTED** [review]
Far too much Python coding, I am afraid - must be done in C++, otherwise we will have to code it all again in Matlab (Octave, Julia, ...). Ideally, all Python interface methods must just prepare and make calls to C interface functions and check their return status.

I would suggest adding `int calibration` argument to `MRAcquisitionData::get_subset` so that for a non-zero value of this argument only the calibration area acqusitions are returned.

**Comment by KrisThielemans** [comment]
> Far too much Python coding, I am afraid - must be done in C++, otherwise we will have to code it all again in Matlab (Octave, Julia, ...). Ideally, all Python interface methods must just prepare and make calls to C interface functions and check their return status.
> 

Pretty much all of the CSM estimation code is Python anyway. So I guess you're mostly/only taking about `extract_calibration_data`? That does seem a prime candidate for a C++ member, but  while I heartily agree with that statement, but I think we have to accept contributions in Python (or MATLAB), and then convert them to C++ in a next PR.

> I would suggest adding `int calibration` argument to `MRAcquisitionData::get_subset` so that for a non-zero value of this argument only the calibration area acqusitions are returned.

I don't think that'd work. `get_subset` returns some k-space lines I guess. Calibration data are just a few of those. I would just add `extract_calibration_data` (later).

**Comment by evgueni-ovtchinnikov** [comment]
@KrisThielemans 

> Pretty much all of the CSM estimation code is Python anyway.

This is largely due to the option to use Inati method from `ismrmrdtools` - without it, Johannes' methods `__calc_from*` would just call respective `cGT_computeCoilSensitivities*` interface functions.

> I don't think that'd work. get_subset returns some k-space lines I guess. 

It would just take adding
```
if (!calibration || acq.isFlagSet(ISMRMRD::ISMRMRD_ACQ_IS_PARALLEL_CALIBRATION) ||
acq.isFlagSet(ISMRMRD::ISMRMRD_ACQ_IS_PARALLEL_CALIBRATION_AND_IMAGING))
```
in front of `subset.append_acquisition(acq);` in `MRAcquisitionData::get_subset`.

**Comment by KrisThielemans** [comment]
> > I don't think that'd work. get_subset returns some k-space lines I guess.
> 
> It would just take adding in `MRAcquisitionData::get_subset`
> ...

I meant that it might not make logical sense, but I'll leave that to @johannesmayer to decide :-)

**Comment by johannesmayer** [comment]
Okay, I agree that @evgueni-ovtchinnikov is right... 
I shifted it all to the C++ level doing:

* `MRAcquisitionData::keep_flagged_acquisitions()` is a method that you can call with a vector of ISMRMRD flags as an argument. It will keep only the acquisitions that have the flag in the container (except if none of them has the flag, then it keeps them all).
* Now only the coil computation code checks for a cartesian trajectory and calls the method with the two calibration flags on the data used for coil sensitivity computation.

File: src/xGadgetron/pGadgetron/Gadgetron.py (hunk 4 above)
Comment (johannesmayer) [reply src/xGadgetron/pGadgetron/Gadgetron.py]: Done

File: src/xGadgetron/pGadgetron/Gadgetron.py (hunk 3 above)
Comment (johannesmayer) [reply src/xGadgetron/pGadgetron/Gadgetron.py]: Done by deletion.

**Pull request merged**